hydra/hydra.go

```diff
@@ -0,0 +1,541 @@
+	"github.com/volatiletech/authboss"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type ConsentValuer interface {
+	authboss.Validator
+
+	GetScopes() []string
+	GetIsAllowed() bool
+	GetRequestedAudience() []string
+}
+
+func MustHaveConsent(v authboss.Validator) ConsentValuer {
+	if u, ok := v.(ConsentValuer); ok {
+		return u
+	}
+
+	panic(fmt.Sprintf("bodyreader returned a type that could not be upgraded to ConsentValuer: %T", v))
+}
+
+type ChallengeValuer interface {
+	authboss.Validator
+
+	GetChallenge() string
+}
+
+func MustHaveChallenge(v authboss.Validator) ChallengeValuer {
+	if u, ok := v.(ChallengeValuer); ok {
+		return u
+	}
+
+	panic(fmt.Sprintf("bodyreader returned a type that could not be upgraded to ChallengeValuer: %T", v))
+}
+
+type LogoutValuer interface {
+	authboss.Validator
+
+	GetShouldLogout() bool
+}
+
+func MustHaveLougout(v authboss.Validator) LogoutValuer {
+	if u, ok := v.(LogoutValuer); ok {
+		return u
+	}
+
+	panic(fmt.Sprintf("bodyreader returned a type that could not be upgraded to LogoutValuer: %T", v))
+}
+
+type SessionableUser interface {
+	authboss.User
+
+	GetSession() (session map[string]interface{})
+}
+
+// TODO: document oauth2 and openID reserved scopes/session keys, potentially type them via an additional module
+const (
+	// PageLogin is for identifying the login page for parsing & validation
+	PageLogin    = "login"
+	PageConsent  = "consent"
+	PageLogout   = "logout"
+	ChallengeKey = "challenge"
+)
+
+func init() {
+	authboss.RegisterModule("hydraconsent", &HydraConsent{})
+}
+
+// HydraConsent module
+type HydraConsent struct {
+	*authboss.Authboss
+	hClient                   *hconsenter.Client
+	rememberMe                int
+	ignoreConsent             map[string]bool
+	overrideRequestedAudience bool
+}
+
+// Init module
+func (a *HydraConsent) Init(ab *authboss.Authboss) (err error) {
+	a.Authboss = ab
+	a.rememberMe = 3600
+
+	if err = a.Authboss.Config.Core.ViewRenderer.Load(PageLogin); err != nil {
+		return err
+	}
+	if err = a.Authboss.Config.Core.ViewRenderer.Load(PageLogout); err != nil {
+		return err
+	}
+	if err = a.Authboss.Config.Core.ViewRenderer.Load(PageConsent); err != nil {
+		return err
+	}
+
+	a.Authboss.Config.Core.Router.Get("/login", a.Authboss.Core.ErrorHandler.Wrap(a.LoginGet))
+	a.Authboss.Config.Core.Router.Post("/login", a.Authboss.Core.ErrorHandler.Wrap(a.LoginPost))
+	a.Authboss.Config.Core.Router.Get("/consent", a.Authboss.Core.ErrorHandler.Wrap(a.ConsentGet))
+	a.Authboss.Config.Core.Router.Post("/consent", a.Authboss.Core.ErrorHandler.Wrap(a.ConsentPost))
+	a.Authboss.Config.Core.Router.Get("/logout", a.Authboss.Core.ErrorHandler.Wrap(a.LogoutGet))
+	a.Authboss.Config.Core.Router.Post("/logout", a.Authboss.Core.ErrorHandler.Wrap(a.LoginPost))
+
+	overrideRaw := os.Getenv("OVERRIDE_REQUESTED_AUDIENCE")
+	override, _ := strconv.ParseBool(overrideRaw)
+	a.overrideRequestedAudience = override
+
+	whiteList := os.Getenv("CONSENT_WHITELIST")
```

Comment:
I'm not sure given hydra or the use case so I'm simply asking the question:
Does it make sense for there to be a check return an error if this is empty?

Reply:
No, the white list is for overriding the consent prompt. Its intended for internal or "1st/primary" applications, ie applications that are trusted automatically with consent. Where as 3rd party applications that want to request auth via hydra should not be trusted and therefore should require the user to submit a consent form for the scopes that are being requested.